README.md

# CPF/CNPJ Validator
  
A Flutter plugin to validate CPF/CNPJ numbers from Brazil.  

[![pub package](https://img.shields.io/pub/v/cpf_cnpj_validator.svg)](https://pub.dartlang.org/packages/cpf_cnpj_validator)
  
## Installation  
Add this to your package's pubspec.yaml file:

```yaml
dependencies:
  cpf_cnpj_validator: 2.0.0-nullsafety
```
  
## Usage CPF  
  
``` dart  
// Import package  
import 'package:cpf_cnpj_validator/cpf_validator.dart';  
  
CPFValidator.isValid("334.616.710-02") // true
CPFValidator.isValid("334.616.710-01") // false
CPFValidator.isValid("35999906032") // true
CPFValidator.isValid("35999906031") // false
CPFValidator.isValid("033461671002") // false

// If you dont want the validation method to strip the values
// Just use false in the second argument
CPFValidator.isValid("334.616.710-02", false) // false
CPFValidator.isValid("35999906032@mail", false) // false


// Other utility methods
CPFValidator.format("33461671002") // Result: 334.616.710-02
CPFValidator.strip("334.616.710-02") // Result: 33461671002

// Generate a raw CPF number, without format
CPFValidator.generate() // Result: 33461671002

// Generate a formatted CPF number
CPFValidator.generate(true) // Result: 334.616.710-02 
```  

## Usage CNPJ  
  
``` dart  
// Import package  
import 'package:cpf_cnpj_validator/cnpj_validator.dart';  
  
CNPJValidator.isValid("12.175.094/0001-19") // true
CNPJValidator.isValid("12.175.094/0001-18") // false
CNPJValidator.isValid("17942159000128") // true
CNPJValidator.isValid("17942159000127") // false
CNPJValidator.isValid("017942159000128") // false

// If you dont want the validation method to strip the values
// Just use false in the second argument
CNPJValidator.isValid("12.175.094/0001-19", false) // false
CNPJValidator.isValid("17942159000128@mail", false) // false

// Other utility methods
CNPJValidator.format("85137090000110") // Result: 85.137.090/0001-10
CNPJValidator.strip("85.137.090/0001-10") // Result: 85137090000110

// Generate a raw CNPJ number, without format
CNPJValidator.generate() // Result: 85137090000110

// Generate a formatted CNPJ number
CNPJValidator.generate(true) // Result: 85.137.090/0001-10 
```  

## Credits

This plugin is a rewrite to Dart from an existing Javascript project with the same purpose.
All the credits go to the person who created the projects in Javascript.

If you guys like this plugin, make sure to star the original projects. I'll leave the links below:

[https://github.com/fnando/cpf](https://github.com/fnando/cpf)

[https://github.com/fnando/cnpj](https://github.com/fnando/cnpj)

## Contributing

If you wish to contribute to this project, I encourage you to open a pull request.

CHANGELOG.md
## v2.1.0

* Remove hard dependency of Flutter.
* Move `test` package dependency to `dev_dependencies`.

## v1.0.0

This release includes:

* Ability to validate CPF and CNPJ numbers.
* Ability to format, strip and create random CPF and CNPJ numbers.

## v1.0.1

* Fixing README.md

## v1.0.2

* Fixing README.md

## v1.0.3

* Fixing README.md

## v1.0.4

* Adding Example
* Changing description

## v1.0.5

* Adding new parameter to control whether or not the validation method should strip the values

## 2.0.0-nullsafety

* Migrating to Null Safety


test/cpf_test.dart
import 'package:cpf_cnpj_validator/cpf_validator.dart';
import 'package:test/test.dart';

void main() {
  test("Test CPF validator", () {
    expect(CPFValidator.isValid("334.616.710-02"), true);
    expect(CPFValidator.isValid("334.616.710-01"), false);
    expect(CPFValidator.isValid("35999906032"), true);
    expect(CPFValidator.isValid("35999906031"), false);
    expect(CPFValidator.isValid("033461671002"), false);
    expect(CPFValidator.isValid("03346teste1671002@mail", false), false);
    expect(CPFValidator.isValid("57abc803.6586-52", false), false);
    expect(CPFValidator.isValid("03.3461.67100-2"), false);

    List<String> blackListed = [
      "00000000000",
      "11111111111",
      "22222222222",
      "33333333333",
      "44444444444",
      "55555555555",
      "66666666666",
      "77777777777",
      "88888888888",
      "99999999999",
      "12345678909"
    ];

    blackListed.forEach((cpf) => expect(CPFValidator.isValid(cpf), false));
  });

  test("Test CPF generator", () {
    for (var i = 0; i < 10; i++) {
      String raw = CPFValidator.generate();
      String formatted = CPFValidator.generate(true);

      expect(raw != formatted, true);
      expect(CPFValidator.isValid(raw), true);
      expect(CPFValidator.isValid(formatted), true);
    }
  });

  test("Test CPF formatter", () {
    expect(CPFValidator.format("33461671002"), "334.616.710-02");
  });

  test("Test CPF strip", () {
    expect(CPFValidator.strip("334.616.710-02"), "33461671002");
  });
}


test/cnpj_test.dart
import 'package:cpf_cnpj_validator/cnpj_validator.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test("Test CNPJ validator", () {
    expect(CNPJValidator.isValid("12.175.094/0001-19"), true);
    expect(CNPJValidator.isValid("12.175.094/0001-18"), false);
    expect(CNPJValidator.isValid("17942159000128"), true);
    expect(CNPJValidator.isValid("17942159000128@mail.com", false), false);
    expect(CNPJValidator.isValid("17942159000128", false), true);
    expect(CNPJValidator.isValid("17942159000127"), false);
    expect(CNPJValidator.isValid("017942159000128"), false);

    List<String> blackListed = [
      "00000000000000",
      "11111111111111",
      "22222222222222",
      "33333333333333",
      "44444444444444",
      "55555555555555",
      "66666666666666",
      "77777777777777",
      "88888888888888",
      "99999999999999"
    ];

    blackListed.forEach((cnpj) => expect(CNPJValidator.isValid(cnpj), false));
  });

  test("Test CNPJ generator", () {
    for (var i = 0; i < 10; i++) {
      String raw = CNPJValidator.generate();
      String formatted = CNPJValidator.generate(true);

      expect(raw != formatted, true);
      expect(CNPJValidator.isValid(raw), true);
      expect(CNPJValidator.isValid(formatted), true);
    }
  });

  test("Test CNPJ formatter", () {
    expect(CNPJValidator.format("85137090000110"), "85.137.090/0001-10");
  });

  test("Test CNPJ strip", () {
    expect(CNPJValidator.strip("85.137.090/0001-10"), "85137090000110");
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:cpf_cnpj_validator/cpf_validator.dart';
import 'package:cpf_cnpj_validator/cnpj_validator.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Text("CPF is valid: ${CPFValidator.isValid("334.616.710-02")}"),
            Text("CNPJ is valid: ${CNPJValidator.isValid("12.175.094/0001-19")}"),
          ],
        ),
      ),
    );
  }
}


lib/cnpj_validator.dart
import 'dart:math';

class CNPJValidator {
  static const List<String> BLACKLIST = [
    "00000000000000",
    "11111111111111",
    "22222222222222",
    "33333333333333",
    "44444444444444",
    "55555555555555",
    "66666666666666",
    "77777777777777",
    "88888888888888",
    "99999999999999"
  ];

  static const STRIP_REGEX = r'[^\d]';

  // Compute the Verifier Digit (or "Dígito Verificador (DV)" in PT-BR).
  // You can learn more about the algorithm on [wikipedia (pt-br)](https://pt.wikipedia.org/wiki/D%C3%ADgito_verificador)
  static int _verifierDigit(String cnpj) {
    int index = 2;

    List<int> reverse =
        cnpj.split("").map((s) => int.parse(s)).toList().reversed.toList();

    int sum = 0;

    reverse.forEach((number) {
      sum += number * index;
      index = (index == 9 ? 2 : index + 1);
    });

    int mod = sum % 11;

    return (mod < 2 ? 0 : 11 - mod);
  }

  static String format(String cnpj) {
    RegExp regExp = RegExp(r'^(\d{2})(\d{3})(\d{3})(\d{4})(\d{2})$');

    return strip(cnpj).replaceAllMapped(
        regExp, (Match m) => "${m[1]}.${m[2]}.${m[3]}/${m[4]}-${m[5]}");
  }

  static String strip(String? cnpj) {
    RegExp regex = RegExp(STRIP_REGEX);
    cnpj = cnpj == null ? "" : cnpj;

    return cnpj.replaceAll(regex, "");
  }

  static bool isValid(String? cnpj, [stripBeforeValidation = true]) {
    if (stripBeforeValidation) {
      cnpj = strip(cnpj);
    }

    // cnpj must be defined
    if (cnpj == null || cnpj.isEmpty) {
      return false;
    }

    // cnpj must have 14 chars
    if (cnpj.length != 14) {
      return false;
    }

    // cnpj can't be blacklisted
    if (BLACKLIST.indexOf(cnpj) != -1) {
      return false;
    }

    String numbers = cnpj.substring(0, 12);
    numbers += _verifierDigit(numbers).toString();
    numbers += _verifierDigit(numbers).toString();

    return numbers.substring(numbers.length - 2) ==
        cnpj.substring(cnpj.length - 2);
  }

  static String generate([bool useFormat = false]) {
    String numbers = "";

    for (var i = 0; i < 12; i += 1) {
      numbers += Random().nextInt(9).toString();
    }

    numbers += _verifierDigit(numbers).toString();
    numbers += _verifierDigit(numbers).toString();

    return (useFormat ? format(numbers) : numbers);
  }
}


lib/cpf_validator.dart
import 'dart:math';

class CPFValidator {
  static const List<String> BLACKLIST = [
    "00000000000",
    "11111111111",
    "22222222222",
    "33333333333",
    "44444444444",
    "55555555555",
    "66666666666",
    "77777777777",
    "88888888888",
    "99999999999",
    "12345678909"
  ];

  static const STRIP_REGEX = r'[^\d]';

  // Compute the Verifier Digit (or "Dígito Verificador (DV)" in PT-BR).
  // You can learn more about the algorithm on [wikipedia (pt-br)](https://pt.wikipedia.org/wiki/D%C3%ADgito_verificador)
  static int _verifierDigit(String cpf) {
    List<int> numbers =
    cpf.split("").map((number) => int.parse(number, radix: 10)).toList();

    int modulus = numbers.length + 1;

    List<int> multiplied = [];

    for (var i = 0; i < numbers.length; i++) {
      multiplied.add(numbers[i] * (modulus - i));
    }

    int mod = multiplied.reduce((buffer, number) => buffer + number) % 11;

    return (mod < 2 ? 0 : 11 - mod);
  }

  static String format(String cpf) {
    RegExp regExp = RegExp(r'^(\d{3})(\d{3})(\d{3})(\d{2})$');

    return strip(cpf).replaceAllMapped(
        regExp, (Match m) => "${m[1]}.${m[2]}.${m[3]}-${m[4]}");
  }

  static String strip(String? cpf) {
    RegExp regExp = RegExp(STRIP_REGEX);
    cpf = cpf == null ? "" : cpf;

    return cpf.replaceAll(regExp, "");
  }

  static bool isValid(String? cpf, [stripBeforeValidation = true]) {
    if (stripBeforeValidation) {
      cpf = strip(cpf);
    }

    // CPF must be defined
    if (cpf == null || cpf.isEmpty) {
      return false;
    }

    // CPF must have 11 chars
    if (cpf.length != 11) {
      return false;
    }

    // CPF can't be blacklisted
    if (BLACKLIST.indexOf(cpf) != -1) {
      return false;
    }

    String numbers = cpf.substring(0, 9);
    numbers += _verifierDigit(numbers).toString();
    numbers += _verifierDigit(numbers).toString();

    return numbers.substring(numbers.length - 2) ==
        cpf.substring(cpf.length - 2);
  }

  static String generate([bool useFormat = false]) {
    String numbers = "";

    for (var i = 0; i < 9; i += 1) {
      numbers += Random().nextInt(9).toString();
    }

    numbers += _verifierDigit(numbers).toString();
    numbers += _verifierDigit(numbers).toString();

    return (useFormat ? format(numbers) : numbers);
  }
}
